django/db/models/fields/json.py

```diff
@@ -0,0 +1,421 @@
+
+    def as_sql_operator(self, compiler, connection, operator):
+        lhs, lhs_params = self.process_lhs(compiler, connection)
+        rhs, rhs_params = self.process_rhs(compiler, connection)
+        params = lhs_params + rhs_params
+        return '%s %s %s' % (lhs, operator, rhs), params
+
+    def as_postgresql(self, compiler, connection):
+        return self.as_sql_operator(compiler, connection, self.postgres_operator)
+
+    def as_sql(self, compiler, connection):
+        raise NotSupportedError(
+            _('%s lookup is not supported by this database backend.' % self.lookup_name)
+        )
+
+
+class HasKeyLookup(SimpleFunctionOperatorMixin, Lookup):
+    logical_operator = None
+
+    def as_sql(self, compiler, connection, template=None):
+        lhs, lhs_params = self.process_lhs(compiler, connection)
+        rhs = [self.rhs] if not isinstance(self.rhs, (list, tuple)) else self.rhs
+        rhs_params = ['$.%s' % json.dumps(key) for key in rhs]
+        sql = template % lhs
+        if self.logical_operator:
+            # Add condition for each key.
+            sql = self.logical_operator.join([sql] * len(rhs_params))
+        return sql, lhs_params + rhs_params
+
+    def as_mysql(self, compiler, connection):
+        return self.as_sql(compiler, connection, template="JSON_CONTAINS_PATH(%s, 'one', %%s)")
+
+    def as_oracle(self, compiler, connection):
+        sql, params = self.as_sql(compiler, connection, template="JSON_EXISTS(%s, '%%s')")
+        # Add paths directly into SQL because path expressions cannot be passed
+        # as bind variables on Oracle.
+        return sql % tuple(params), []
+
+    def as_sqlite(self, compiler, connection):
+        return self.as_sql(compiler, connection, template='JSON_TYPE(%s, %%s) IS NOT NULL')
+
+
+@JSONField.register_lookup
+class HasKey(HasKeyLookup):
+    lookup_name = 'has_key'
+    postgres_operator = '?'
+
+    prepare_rhs = False
+
+
+@JSONField.register_lookup
+class HasAnyKeys(HasKeyLookup):
+    lookup_name = 'has_any_keys'
+    postgres_operator = '?|'
+    logical_operator = ' OR '
+
+    def get_prep_lookup(self):
+        return [str(item) for item in self.rhs]
+
+
+@JSONField.register_lookup
+class HasKeys(HasAnyKeys):
+    lookup_name = 'has_keys'
+    postgres_operator = '?&'
+    logical_operator = ' AND '
+
+
+@JSONField.register_lookup
+class DataContains(SimpleFunctionOperatorMixin, Lookup):
+    lookup_name = 'contains'
+    postgres_operator = '@>'
+
+    def as_mysql(self, compiler, connection, flipped=False):
+        return super().as_sql_function(
+            compiler, connection, template="JSON_CONTAINS(%s, %s, '$')", flipped=flipped
+        )
+
+
+@JSONField.register_lookup
+class ContainedBy(DataContains):
+    lookup_name = 'contained_by'
+    postgres_operator = '<@'
+
+    def as_mysql(self, compiler, connection):
+        return super().as_mysql(compiler, connection, flipped=True)
+
+
+class JSONValue(Func):
```

Comment:
I think you can use `django.db.models.function.Cast` instead:

```
class JSONValue(Cast):
    def __init__(self, expression):
        super().__init__(expression=Value(expression), output_field=JSONField)
```

Reply:
Well, to be honest, I don't know if I should include that in the code since the current implementation doesn't use that function anymore. Also, `CAST AS JSON` only works on MySQL. For other backends, the function would need to use other JSON functions.

Reply:
I guess you could implement `as_postgresql()` here and use `::jsonb`?

Reply:
@pope1ni The `::type` happens with `Cast`.

Reply:
I decided to remove this because it doesn't really make sense. Casting to JSON on database backends that have native JSON data type can already be done using `Cast` like other fields, e.g. `Cast(expr, JSONField())`. I initially used the function (as copied from [django-jsonfallback](https://github.com/raphaelm/django-jsonfallback/blob/master/jsonfallback/fields.py#L242)), but now the implementation doesn't need the function anymore.